x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/AllSupportedFieldsTestCase.java

```diff
@@ -0,0 +1,429 @@
+        };
+    }
+
+    /**
+     * Is the type supported in indices?
+     */
+    private static boolean supportedInIndex(DataType t) {
+        return switch (t) {
+            // These are supported but implied by the index process.
+            case OBJECT, SOURCE, DOC_DATA_TYPE, TSID_DATA_TYPE,
+                // Internal only
+                UNSUPPORTED, PARTIAL_AGG,
+                // You can't index these - they are just constants.
+                DATE_PERIOD, TIME_DURATION, GEOTILE, GEOHASH, GEOHEX,
+                // TODO fix geo
+                CARTESIAN_POINT, CARTESIAN_SHAPE -> false;
+            default -> true;
+        };
+    }
+
+    private Map<String, Object> nameToType(List<?> columns) {
+        Map<String, Object> result = new TreeMap<>();
+        for (Object c : columns) {
+            Map<?, ?> map = (Map<?, ?>) c;
+            result.put(map.get("name").toString(), map.get("type"));
+        }
+        return result;
+    }
+
+    private List<String> names(List<?> columns) {
+        List<String> result = new ArrayList<>();
+        for (Object c : columns) {
+            Map<?, ?> map = (Map<?, ?>) c;
+            result.add(map.get("name").toString());
+        }
+        return result;
+    }
+
+    private Map<String, Map<String, Object>> indexToRow(List<?> columns, List<?> values) {
+        List<String> names = names(columns);
+        int timestampIdx = names.indexOf("_index");
+        Map<String, Map<String, Object>> result = new TreeMap<>();
+        for (Object r : values) {
+            List<?> row = (List<?>) r;
+            result.put(row.get(timestampIdx).toString(), nameToValue(names, row));
+        }
+        return result;
+    }
+
+    private Map<String, Object> nameToValue(List<String> names, List<?> values) {
+        Map<String, Object> result = new TreeMap<>();
+        for (int i = 0; i < values.size(); i++) {
+            result.put(names.get(i), values.get(i));
+        }
+        return result;
+    }
+
+    private Matcher<String> expectedType(Version minVersion, DataType type) {
+        return switch (type) {
+            case COUNTER_DOUBLE, COUNTER_LONG, COUNTER_INTEGER -> {
+                if (indexMode == IndexMode.TIME_SERIES) {
+                    yield equalTo(type.esType());
+                }
+                yield equalTo(type.esType().replace("counter_", ""));
+            }
+            case BYTE, SHORT -> equalTo("integer");
+            case HALF_FLOAT, SCALED_FLOAT, FLOAT -> equalTo("double");
+            case NULL -> equalTo("keyword");
+            // Currently unsupported without TS command or KNN function
+            case AGGREGATE_METRIC_DOUBLE, DENSE_VECTOR -> {
+                if (false == minVersion.onOrAfter(Version.V_9_2_0)) {
+                    yield either(equalTo("unsupported")).or(equalTo(type.esType()));
+                }
+                yield equalTo("unsupported");
```

Comment:
Should we also test that the `originalType` is correctly reported in the output? That's something we could mess up when tinkering with the under construction types.